### ISSUE TITLE
Webpack CommonsChunkPlugin assumes output files will only be run in the main thread.

### ISSUE THREAD

**Issue by jshcrowthe** [webpack/webpack#4819]
**Do you want to request a *feature* or report a *bug*?**
bug

**What is the current behavior?**

Webpack, when using the `webpack.optimize.CommonsChunkPlugin`, assumes that the output files will only be run in the main browser thread and attaches the `webpackJsonp` function to the `window` object.

**If the current behavior is a bug, please provide the steps to reproduce.**

Repo outlining the issue here: https://github.com/jshcrowthe/webpack-scoping-issue.

Running version available here: https://jshcrowthe.github.io/webpack-scoping-issue

1. Open the link above in a browser that supports service worker
1. Open the console
1. Observe the `window is not defined` error (If you see a `ServiceWorker script evaluation failed` but not the error mentioned, refresh and you should see the error)

**What is the expected behavior?**

Webpack should grab the global scope for the current thread and attach the `webpackJsonp` function there.

**If this is a feature request, what is motivation or use case for changing the behavior?**

Chunked libraries can not be run in any thread w/o a `window` object as its global.

**Please mention other relevant information such as the browser version, Node.js version, webpack version and Operating System.**

Node.js: `v6.10.1`
Webpack `v2.4.1`
OS: macOS Sierra 10.12.4

**Comment by sokra** [comment]
There is a `target` option that allows to compile for WebWorker.

**Comment by jshcrowthe** [comment]
Thanks for the response @sokra! I didn't know about the `target` options (sorry, my ignorance is showing). 

So I took a jab at it, and it seems that specifying the `webworker` target prevents the script from running in the main thread due to its use of `importScripts` (see: [this commit](https://github.com/jshcrowthe/webpack-scoping-issue/blob/webworker-target/dist/lib.js#L43-L51)). In addition the built file does not seem to run properly in either the service worker or the main thread (again, I may just be ignorant).

Any thoughts?

**Comment by TheLarkInn** [comment]
@jshcrowthe this isn't entirely relevant to thread but I was curious if this was a G project or personal? Nice to see some Google SWEs in issues and making contributions an was really just curious. 

Also I'm getting around to #5530. Would this help solve your case here?

**Comment by jshcrowthe** [comment]
This was for the [Firebase JS SDK](https://github.com/firebase/firebase-js-sdk), we have some pieces of the SDK that we load in both a worker and a browser thread (i.e. the main firebase bootstrapping logic) and dynamically evaluating the global would have solved our use case.

Interestingly enough #5530 is a dupe of the PR I created originally #4820. However I closed it because the solution of computing the global scope doesn't extend to handling the lazily loaded chunks which _is_ thread specific. 

_e.g._

https://github.com/jshcrowthe/webpack/blob/ce036c5ab3a32b8deb6ae153865d344c5a50209f/lib/JsonpMainTemplatePlugin.js#L49-L78

This comment by @johnthecat was something I didn't realize (I'm still learning the webpack-foo) but once I saw it agreed with his solution:

> I think this change is redundant, because it still loads chunks by creating script tag.
Some time ago I was forced to write my own template plugin, which supports both "web" and "webworker" environments. In my opinion, it would be great to merge JsonpTemplatePlugin and WebWorkerTemplatePlugin into one and use it in most cases. Node environment is more specific and need to have own template plugin, so checks like if (typeof global !== 'undefined') {} are unusable here.

I haven't taken the time to learn how I would go about that though and we worked around the original issue. However w/o that piece, dynamically evaluating the global is only half the solution, even if it solves the immediate problem.
